### PR TITLE
Add support to manage the vlan protocol on a vf

### PR DESCRIPTION
**Pull request by mlguerrero12**
These changes allow users to get and set the vlan protocol on a vf.

### CODE REVIEW

**Comment by mlguerrero12** [comment]
@vishvananda , @aboch , could you please take a look at this?

This will be needed for a new feature in the sriov cni.

**Comment by mlguerrero12** [comment]
Hi @aboch, please let me know if there is something that I can do to speed up the review of this PR? Thank you.

**Comment by cgoncalves** [comment]
LGTM

**Review by SchSeba — APPROVED** [review]
Nice work!

**Comment by mlguerrero12** [comment]
Hi @aboch, any chance to get a review? We want to add vlan protocol support in the sriov cni. This is needed in https://github.com/k8snetworkplumbingwg/sriov-cni/pull/277.

**Comment by mlguerrero12** [comment]
@aboch, @vishvananda, any chance to get a review on this please?

File: link_linux.go [hunk 1]
```diff
@@ -612,6 +612,41 @@ func (h *Handle) LinkSetVfVlanQos(link Link, vf, vlan, qos int) error {
...
+	data := nl.NewRtAttr(unix.IFLA_VFINFO_LIST, nil)
+	vfInfo := data.AddRtAttr(nl.IFLA_VF_INFO, nil)
+	vfVlanList := vfInfo.AddRtAttr(nl.IFLA_VF_VLAN_LIST, nil)
+
+	vfmsg := nl.VfVlanInfo{VfVlan: nl.VfVlan{
+		Vf:   uint32(vf),
+		Vlan: uint32(vlan),
+		Qos:  uint32(qos),
+	},
+		VlanProto: (uint16(proto)>>8)&0xFF | (uint16(proto)&0xFF)<<8}
```
Comment (aboch) [inline link_linux.go:641]: pls fix this formatting

File: link_linux.go [hunk 2]
```diff
@@ -612,6 +612,41 @@ func (h *Handle) LinkSetVfVlanQos(link Link, vf, vlan, qos int) error {
 	return err
 }
 
+// LinkSetVfVlanQosProto sets the vlan, qos and protocol of a vf for the link.
+// Equivalent to: `ip link set $link vf $vf vlan $vlan qos $qos proto $proto`
+func LinkSetVfVlanQosProto(link Link, vf, vlan, qos, proto int) error {
```
Comment (aboch) [inline link_linux.go:617]: Given this takes already 5 parameters and who knows if more are to come in future, would it make sense to pass a structure as input parameter?

File: link_linux.go (hunk 1 above)
Comment (mlguerrero12) [reply link_linux.go]: done, thanks.

File: link_linux.go (hunk 2 above)
Comment (mlguerrero12) [reply link_linux.go:617]: Well, I believe that it is better to have the parameters separated to be consistent with the previous 2 functions (LinkSetVfVlanQos and LinkSetVfVlan). We could have used a structure as input parameter if we would have created a more general function for all vlan options. Currently, the name of the function has a one-to-one relationship with the parameters.

**Comment by mlguerrero12** [comment]
Thanks @aboch for the comments. Please have a look again.

**Comment by aboch** [comment]
Is it possible to add a UT for this?
UT is is about programming the kernel and then read from the kernel the expected parameters

**Comment by mlguerrero12** [comment]
We would need a nic that supports sriov with some VFs configured. I suppose that's the reason none of the VF related functions (txrate, state, spoofchk, vlan, vlanqos) have unit tests. I can only tell you that I fully tested this functionality for a custom sriov cni image that was provided to one of our customers as tech preview. No complaints so far. @aboch

**Comment by aboch** [comment]
Unless it is really two separate functionality changes, please squash the two commits into one.
This project follows a "one commit per PR" idea.

**Comment by mlguerrero12** [comment]
@aboch, done. Thanks.

**Pull request merged**